README.org

```diff
@@ -7,8 +7,33 @@ s-expressions.  S-expressions are defined by the following type:
 type sexp = Atom of string | List of sexp list
 #+end_src
 
-and are rendered as parenthesized lists of strings, e.g., =(This (is
-an) (s expression))=.
+
+** Usage example
+
+In this example we build an Sexp expression that corresponds to
+=(This (is an) (s expression))=, serialize a Sexp expression into a string,
+and then parse a string into an Sexp.
+
+#+begin_src ocaml
+open Sexplib
+
+let () =
+  (* Build an Sexp from: (This (is an) (s expression)) *)
+  let exp1 = Sexp.(List [
+    Atom "This";
+    List [Atom "is"; Atom "an"];
+    List [Atom "s"; Atom "expression"]
+  ]) in
+  (* Serialize an Sexp object into a string *)
+  print_endline (Sexp.to_string exp1);
+  (* Parse a string and produce a Sexp object  *)
+  let exp2 = Sexp.of_string "(This (is an) (s expression))" in
+  (* Ensure we parsed what we expected. *)
+  assert (exp1 = exp2)
```

Comment:
Better to use Sexp.equal or Sexp.(=), rather than polymorphic equality.

Reply:
I am unable to get that change to compile.
When I do
```
  assert (Sexp.(=) exp1 exp2)
```
I get the following error:
```
File "sexplib-hello-world/main.ml", line 15, characters 10-18:
Error: Unbound value Sexp.=
```

I did look at [1] but couldn't find `=` defined there. What am I doing wrong?

1. https://github.com/janestreet/sexplib/blob/master/src/sexp_intf.ml

Reply:
Huh, I'm surprised that's not there. You can always do [%compare.equal: Sexp.t], though that depends on ppx_compare.

Reply:
In order to compare two sexp values we need to load another model?

Can you clarify in which situations the regular polymorphic fails?

Reply:
Polymorphic compare will never fail on a Sexp.t. But it's slower than a code-generated comparison function, and polymorphic compare can in many cases do crazy things. For example, you can generate sets (say, of the Set.Tree.t type) where Set.Tree.(=) x y evaluates to true, but x = y evaluates to false. That's because polymorphic compare doesn't know that binary trees that have different shape may represent the same underlying set.

Here's an old blog post about it: https://blog.janestreet.com/the-perils-of-polymorphic-compare/

Reply:
Another alternative is to do [Sexp.compare x y = 0], which is the same thing that [%compare.equal: Sexp.t] generates.

Reply:
If the polymorphic equals will never fail on a Sexp.t and the point of the text is on showing a usage example of Sexpt.t, then why do the other two arguments matter? Both arguments, while important as general observations on OCaml, seem a bit tangential to the point of the *usage*.  

In this example the assertion is simply stating that the two values have the same structure, which is an important observation of the parser. The performance consideration of compare versus equal seems to be a bit of minutia (which may be observed in the usage example). 

Regardless, if you remain unconvinced and would rather accept the patch with `Sexp.compre x y = 0`, I'll go head and change it.

Reply:
Yeah, I prefer not to use polymorphic compare at all. So, imperfect though it is, I would prefer that. (FWIW, if you have Base open, polymorphic compare is hidden by default, which is another reason not to express the example this way.)